### ISSUE TITLE
Bug: [only-throw-error] TypeOrValueSpecifier does not handle union

### ISSUE THREAD

**Issue by BrendanC23** [typescript-eslint/typescript-eslint#11452]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.5&fileType=.ts&code=GYVwdgxgLglg9mABMAjAHgCoD4AUBKRAbwChFEAbAUykUoCc6AuRABTrgFsYBnSzLANylEUABbsA7rQZCAvsWKhIsBMgBMmFABpEGNbgIkyVGvSat2XXpqyIAPhc48%2BewcLGTpdOUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AeyfkNvwAto7AO71og6OjABtcNhyIx7CQBpZcrHOyQAhvHjDJMzZsgAFQQFtKyRJDWaAuvYgBfe09cgXQA&tsconfig=&tokens=false

### Repro Code

```TypeScript
function f1<T>() {
  let err: Promise<T>;
  throw err;
}

function f2<T1, T2>() {
  let err: Promise<T1> | Promise<T2>;
  throw err;
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/only-throw-error": [
      "error",
      {
        "allow": [
          "Promise"
        ]
      }
    ]
  }
}
```

### tsconfig

```jsonc

```

### Expected Result

The only-throw-error warning should not be triggered by `f1()` or `f2()`. When the "allow" option is `Promise`, I should be able to throw an error of `Promise<A> | Promise<B>`.

### Actual Result

The "only-throw-error" rule triggers when a union is thrown.

> Expected an error object to be thrown. 8:9 - 8:12

<img width="727" height="247" alt="Expected an error object to be thrown. 8:9 - 8:12" src="https://github.com/user-attachments/assets/2c38a7ad-911e-43f2-8c30-4510ef995efc" />

### Additional Info

My actual use-case is with TanStack Router, which allows users to `throw redirect(...)` (this throws a `Redirect` type). I have a function that can return various redirects (which are later thrown):

```ts
function getRedirect(a, b) {
    if (a) return redirect(...); // returns Redirect<A>
    else return redirect(...); // returns Redirect<B>
}
```

This returns a union of `Redirect<A> | Redirect<B>`. The following code triggers and ESLint error because it is trying to throw a union type.

```ts
throw getRedirect(a, b);
```

**Comment by JoshuaKGoldberg** [comment]
Nifty: `Promise<T> | Promise<T>` and `Promise<T1> | Promise<T1>` don't reproduce the issue. TypeScript is collapsing those `===` types into one.

This feels like a conceptually direct bug to me: right now, if every constituent of a union type is in the list, the union type should be matched. Right now it's not.

Accepting PRs - thanks for filing!
